# Dask

Dask ist mein persönlicher Favorit zur Arbeit mit Datenmengen größer Arbeitsspeicher. 
Es gibt in Dask mehrere Sub-Packages zur Arbeit mit verschiedenen Datensets. Dataframes, Arrays und JSON Dokumente können mit Dask prozessiert werden<br>
<a href='https://dask.pydata.org/en/latest/'> Dask API </a>

Woher bekomme ich Dask? 
http://dask.pydata.org/en/latest/install.html
oder pip install dask/ condas install dask

Dask bietet DataFrames im bekannten Pandas Format. Die Abbildung zeigt wie Dask arbeitet. Es zerlegt die eingelesene Datei mehrere Subsampels/DataFrames und bearbeitet jedes mal ein Dataframe im Arbeitsspeicher.<br>
<img src='dask-dataframe.png'>

Dask Array dem Numpy Array nachempfunden. Es funktioniert in mehrere Dimensionen und nicht nur in die Länge wie das DataFrame.<br>
<img src='dask-array-black-text.png'>

Für JSON und unstrukturierte Daten kann man Dask Bag verwenden.

Wie funktioniert Dask generell?<br>
Dask zieht erstmal nur einen kleinen Ausschnitt der Daten um einen Überblick für den User zu erstellen.
Man kann sich immer einen Ausschnitt aus den oberen Spalten DataFram.head() geben lassen, wird aber nie das gesamte DataFrame sehen.
Im Beispiel werden wir das Taxiset mit 100.000 Reihen von vorher nutzen.
Kann aber mit wesentlich größeren Daten Arbeiten.<br>
Wer es mal testen will <a href='https://data.cityofnewyork.us/Transportation/2015-Yellow-Taxi-Trip-Data/ba8s-jw6u'> NYC Open Data Taxi</a>
160 Millionen Zeilen, 10 GB waren ohne Probleme zu bearbeiten


In [1]:
import dask.dataframe as dd
df = dd.read_csv(r"taxi_hunderttausend.csv",error_bad_lines=False)
df

Dask DataFrame Structure:
                 No. vendorid pickup_datetime dropoff_datetime Store_and_fwd_flag rate_code Pickup_longitude Pickup_latitude Dropoff_longitude Dropoff_latitude Passenger_count Trip_distance Fare_amount    Extra  MTA_tax Tip_amount Tolls_amount Ehail_fee Improvement_surcharge Total_amount Payment_type Trip_type
npartitions=1                                                                                                                                                                                                                                                                                                          
None           int64    int64          object           object             object     int64          float64         float64           float64          float64           int64       float64     float64  float64  float64    float64      float64   float64               float64      float64        int64     int64
None             ...      ...             ...              ...                ...       ...              ...             ...               ...              ...             ...           ...         ...      ...      ...        ...          ...       ...                   ...          ...          ...       ...
Dask Name: from-delayed, 3 tasks

Die Datei könnte ja zu groß seien um Dataframes zu visualisieren. Deshalb immer der Befehl head() um die oberen Spalten anzuzeigen

In [2]:
col = df.columns
col

Index(['No.', 'vendorid', 'pickup_datetime', 'dropoff_datetime',
       'Store_and_fwd_flag', 'rate_code', 'Pickup_longitude',
       'Pickup_latitude', 'Dropoff_longitude', 'Dropoff_latitude',
       'Passenger_count', 'Trip_distance', 'Fare_amount', 'Extra', 'MTA_tax',
       'Tip_amount', 'Tolls_amount', 'Ehail_fee', 'Improvement_surcharge',
       'Total_amount', 'Payment_type', 'Trip_type'],
      dtype='object')

In [3]:
df.head(10)

No.  vendorid         pickup_datetime        dropoff_datetime  \
0    1         1  03/12/2015 09:26:05 AM  03/12/2015 09:33:58 AM   
1    2         1  05/29/2015 03:27:07 AM  05/29/2015 03:45:28 AM   
2    3         2  11/24/2015 08:07:46 PM  11/24/2015 08:31:35 PM   
3    4         2  12/18/2015 06:46:25 PM  12/18/2015 06:53:06 PM   
4    5         2  09/22/2015 06:32:36 PM  09/22/2015 06:34:44 PM   
5    6         2  05/15/2015 08:42:44 PM  05/15/2015 08:56:03 PM   
6    7         2  09/02/2015 02:38:54 PM  09/02/2015 02:45:25 PM   
7    8         2  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8    9         2  05/30/2015 04:22:47 AM  05/30/2015 04:42:57 AM   
9   10         2  03/29/2015 01:46:20 PM  03/29/2015 01:56:56 PM   

  Store_and_fwd_flag  rate_code  Pickup_longitude  Pickup_latitude  \
0                  N          1        -73.950478        40.826733   
1                  N          1        -73.914124        40.823772   
2                  N          1        -73.952751        40.727180   
3                  N          1        -73.960274        40.720379   
4                  N          1        -73.829956        40.713699   
5                  N          1        -73.964096        40.710148   
6                  N          1        -73.830490        40.759331   
7                  N          1        -73.955406        40.721008   
8                  N          1        -73.956276        40.721710   
9                  N          1        -73.954079        40.811382   

   Dropoff_longitude  Dropoff_latitude    ...      Fare_amount  Extra  \
0         -73.971558         40.795403    ...             10.0    0.0   
1         -73.975517         40.751713    ...             23.0    0.5   
2         -73.971313         40.693260    ...             19.5    0.5   
3         -73.956093         40.732990    ...              6.5    1.0   
4         -73.823364         40.713306    ...              4.0    1.0   
5         -73.994087         40.695229    ...             13.5    0.5   
6         -73.843788         40.771461    ...              7.0    0.0   
7         -74.005081         40.739941    ...             29.0    0.5   
8         -73.906715         40.694427    ...             16.0    0.5   
9         -73.945732         40.833076    ...              9.0    0.0   

   MTA_tax  Tip_amount  Tolls_amount  Ehail_fee  Improvement_surcharge  \
0      0.5        1.00           0.0        NaN                    0.3   
1      0.5        0.00           0.0        NaN                    0.3   
2      0.5        0.00           0.0        NaN                    0.3   
3      0.5        2.00           0.0        NaN                    0.3   
4      0.5        0.00           0.0        NaN                    0.3   
5      0.5        0.00           0.0        NaN                    0.3   
6      0.5        1.56           0.0        NaN                    0.3   
7      0.5        6.06           0.0        NaN                    0.3   
8      0.5        3.46           0.0        NaN                    0.3   
9      0.5        0.00           0.0        NaN                    0.3   

   Total_amount  Payment_type  Trip_type  
0         11.80             1          1  
1         24.30             2          1  
2         20.80             2          1  
3         10.30             1          1  
4          5.80             2          1  
5         14.80             2          1  
6          9.36             1          1  
7         36.36             1          1  
8         20.76             1          1  
9          9.80             2          1  

[10 rows x 22 columns]

In [4]:
import sys
objectsize = sys.getsizeof(df)
objectsize

56

Das Dask.DataFrame ist nur 56-Bit groß. Es enthält keine Daten wie der Befehl df oben zeigt. Es ist lediglich ein Verweis über Struktur und Ort der Datei

In [5]:
df = df[df['Total_amount']>15]

In [6]:
df.head(10)

No.  vendorid         pickup_datetime        dropoff_datetime  \
1     2         1  05/29/2015 03:27:07 AM  05/29/2015 03:45:28 AM   
2     3         2  11/24/2015 08:07:46 PM  11/24/2015 08:31:35 PM   
7     8         2  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8     9         2  05/30/2015 04:22:47 AM  05/30/2015 04:42:57 AM   
10   11         2  08/30/2015 09:31:46 PM  08/30/2015 10:04:30 PM   
15   16         2  06/02/2015 11:50:25 AM  06/02/2015 12:15:20 PM   
19   20         2  03/27/2015 12:57:02 AM  03/27/2015 01:12:57 AM   
23   24         2  07/27/2015 02:21:58 PM  07/27/2015 02:33:22 PM   
24   25         2  03/11/2015 11:48:20 AM  03/11/2015 12:02:10 PM   
29   30         2  06/12/2015 06:17:40 PM  06/12/2015 06:32:03 PM   

   Store_and_fwd_flag  rate_code  Pickup_longitude  Pickup_latitude  \
1                   N          1        -73.914124        40.823772   
2                   N          1        -73.952751        40.727180   
7                   N          1        -73.955406        40.721008   
8                   N          1        -73.956276        40.721710   
10                  N          1        -73.844215        40.721420   
15                  N          1        -73.956245        40.747505   
19                  N          1        -73.955009        40.722370   
23                  N          1        -73.950211        40.827045   
24                  N          1        -73.966980        40.705830   
29                  N          1        -73.860786        40.833641   

    Dropoff_longitude  Dropoff_latitude    ...      Fare_amount  Extra  \
1          -73.975517         40.751713    ...             23.0    0.5   
2          -73.971313         40.693260    ...             19.5    0.5   
7          -74.005081         40.739941    ...             29.0    0.5   
8          -73.906715         40.694427    ...             16.0    0.5   
10         -73.858116         40.723831    ...             24.5    0.5   
15         -73.969193         40.750237    ...             18.0    0.0   
19         -73.968163         40.684406    ...             14.0    0.5   
23         -73.979424         40.789413    ...             12.0    0.0   
24         -73.978256         40.686993    ...             12.0    0.0   
29         -73.928329         40.808998    ...             16.5    1.0   

    MTA_tax  Tip_amount  Tolls_amount  Ehail_fee  Improvement_surcharge  \
1       0.5        0.00           0.0        NaN                    0.3   
2       0.5        0.00           0.0        NaN                    0.3   
7       0.5        6.06           0.0        NaN                    0.3   
8       0.5        3.46           0.0        NaN                    0.3   
10      0.5        0.00           0.0        NaN                    0.3   
15      0.5        0.00           0.0        NaN                    0.3   
19      0.5        3.06           0.0        NaN                    0.3   
23      0.5        2.56           0.0        NaN                    0.3   
24      0.5        2.56           0.0        NaN                    0.3   
29      0.5        0.00           0.0        NaN                    0.3   

    Total_amount  Payment_type  Trip_type  
1          24.30             2          1  
2          20.80             2          1  
7          36.36             1          1  
8          20.76             1          1  
10         25.80             2          1  
15         18.80             2          1  
19         18.36             1          1  
23         15.36             1          1  
24         15.36             1          1  
29         18.30             2          1  

[10 rows x 22 columns]

berechnen wir mal den Durchschnittlichen Fahrpreis

In [8]:
x = df.Total_amount.mean()
x

dd.Scalar<series-..., dtype=float64>

Warum ist mean() nur ein Skalar?
df ist immernoch lediglich ein Verweis und mean() ist genauso nur ein Befehl/Verweis. Um aus den Befehlen und Strukturen jetzt einen realen Wert zu berechnen fehlt, eine Aufforderung zum ausführen.<br>
Dieser Befehl lautet compute()


In [9]:
x.compute()

26.10853930217154

Aus den Verweisen wurde ein realer Wert

# Wie wirkt sich compute() auf andere Operationen aus?

Beispiel wir erzeugen df>15 schon direkt als Dataframe beim erzeugen

In [10]:
df = dd.read_csv(r"taxi_hunderttausend.csv",error_bad_lines=False)
df = df[df['Total_amount']>15].compute()
df.head(10)

No.  vendorid         pickup_datetime        dropoff_datetime  \
1     2         1  05/29/2015 03:27:07 AM  05/29/2015 03:45:28 AM   
2     3         2  11/24/2015 08:07:46 PM  11/24/2015 08:31:35 PM   
7     8         2  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8     9         2  05/30/2015 04:22:47 AM  05/30/2015 04:42:57 AM   
10   11         2  08/30/2015 09:31:46 PM  08/30/2015 10:04:30 PM   
15   16         2  06/02/2015 11:50:25 AM  06/02/2015 12:15:20 PM   
19   20         2  03/27/2015 12:57:02 AM  03/27/2015 01:12:57 AM   
23   24         2  07/27/2015 02:21:58 PM  07/27/2015 02:33:22 PM   
24   25         2  03/11/2015 11:48:20 AM  03/11/2015 12:02:10 PM   
29   30         2  06/12/2015 06:17:40 PM  06/12/2015 06:32:03 PM   

   Store_and_fwd_flag  rate_code  Pickup_longitude  Pickup_latitude  \
1                   N          1        -73.914124        40.823772   
2                   N          1        -73.952751        40.727180   
7                   N          1        -73.955406        40.721008   
8                   N          1        -73.956276        40.721710   
10                  N          1        -73.844215        40.721420   
15                  N          1        -73.956245        40.747505   
19                  N          1        -73.955009        40.722370   
23                  N          1        -73.950211        40.827045   
24                  N          1        -73.966980        40.705830   
29                  N          1        -73.860786        40.833641   

    Dropoff_longitude  Dropoff_latitude    ...      Fare_amount  Extra  \
1          -73.975517         40.751713    ...             23.0    0.5   
2          -73.971313         40.693260    ...             19.5    0.5   
7          -74.005081         40.739941    ...             29.0    0.5   
8          -73.906715         40.694427    ...             16.0    0.5   
10         -73.858116         40.723831    ...             24.5    0.5   
15         -73.969193         40.750237    ...             18.0    0.0   
19         -73.968163         40.684406    ...             14.0    0.5   
23         -73.979424         40.789413    ...             12.0    0.0   
24         -73.978256         40.686993    ...             12.0    0.0   
29         -73.928329         40.808998    ...             16.5    1.0   

    MTA_tax  Tip_amount  Tolls_amount  Ehail_fee  Improvement_surcharge  \
1       0.5        0.00           0.0        NaN                    0.3   
2       0.5        0.00           0.0        NaN                    0.3   
7       0.5        6.06           0.0        NaN                    0.3   
8       0.5        3.46           0.0        NaN                    0.3   
10      0.5        0.00           0.0        NaN                    0.3   
15      0.5        0.00           0.0        NaN                    0.3   
19      0.5        3.06           0.0        NaN                    0.3   
23      0.5        2.56           0.0        NaN                    0.3   
24      0.5        2.56           0.0        NaN                    0.3   
29      0.5        0.00           0.0        NaN                    0.3   

    Total_amount  Payment_type  Trip_type  
1          24.30             2          1  
2          20.80             2          1  
7          36.36             1          1  
8          20.76             1          1  
10         25.80             2          1  
15         18.80             2          1  
19         18.36             1          1  
23         15.36             1          1  
24         15.36             1          1  
29         18.30             2          1  

[10 rows x 22 columns]

In [11]:
objectsize = sys.getsizeof(df)
objectsize

12622484

In [12]:
x = df.Total_amount.mean()
x

26.10853930217154

# Wieso existiert compute() überhaupt?

Man kann mehrere Befehle hintereinander schalten ohne die Berechnung die bei 10GB extrem lange dauern kann auszuführen.
Wenn man statt mean() , maen() ausversehen schreibt und erst df>15 ausgeführt wird gibt es einen Error nachdem er das Dataframe erzeugt hat. 

In [13]:
df = dd.read_csv(r"taxi_hunderttausend.csv",error_bad_lines=False)
df = df[df['Total_amount']>15]
y = df.maen()
y.compute()

AttributeError: 'DataFrame' object has no attribute 'maen'

df ist immernoch ein Verweis, aber er erzeugt einen Fehler ohne df vorher berechnet zu haben.

Um das DataFrame in eine CSV Datei zu verwandeln benötigen wir den Befehl df.to_csv. Der Dateiname benötigt noch ein .*. um die verschiedenen Partitionen abzuspeichern

In [14]:
df = dd.read_csv(r"taxi_hunderttausend.csv",error_bad_lines=False)

# Speichern eines Dask Dataframes

In [15]:
df.to_csv(r"taxipartition.*.csv",compression = 'xz',mode = 'a')

man kann die Datei auch komprimieren'compression = 'compression der Wahl'' und mode bestimmt den Umgang wenn eine Datei mit selben Namen bereits vorhanden ist. 
    r for reading,
    w for writing,
    r+ opens for reading and writing (cannot truncate a file),
    w+ for writing and reading (can truncate a file),
    rb+ reading or writing a binary file,
    wb+ writing a binary file,
    a opens for appending,
    x open for exclusive creation, failing if the file already exists (Python 3)
<img src='compression.png'>
und so sieht die Datei in Excel nach der compression aus
<img src='compression2.png'>

So sieht der Ordner aus wenn man größere Dateien verwendet und speichert.<br>
Die Datei ist jetzt in Partitionen zerlegt und als ganze Datei vorhanden. Nicht vergessen, kostet eine Menge an Speicher<br>
Dask hat kein Problem damit die einzelnen Partitionen als ganzes zu laden

<img src='partition.png'>

Wie kann ich jetzt mit den einzelnen Partitionen weiterarbeiten?
Einfach den Zielordner angeben und .001,.002 ... durch .*. ersetzen

In [16]:
df2 = dd.read_csv(r"taxipartition.*.csv",compression = 'xz')
df2.head(10)

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\csv.py:209: UserWarning: Warning xz compression does not support breaking apart files
Please ensure that each individual file can fit in memory and
use the keyword ``blocksize=None to remove this message``
Setting ``blocksize=None``
  "Setting ``blocksize=None``" % compression)


Unnamed: 0  No.  vendorid         pickup_datetime        dropoff_datetime  \
0           0    1         1  03/12/2015 09:26:05 AM  03/12/2015 09:33:58 AM   
1           1    2         1  05/29/2015 03:27:07 AM  05/29/2015 03:45:28 AM   
2           2    3         2  11/24/2015 08:07:46 PM  11/24/2015 08:31:35 PM   
3           3    4         2  12/18/2015 06:46:25 PM  12/18/2015 06:53:06 PM   
4           4    5         2  09/22/2015 06:32:36 PM  09/22/2015 06:34:44 PM   
5           5    6         2  05/15/2015 08:42:44 PM  05/15/2015 08:56:03 PM   
6           6    7         2  09/02/2015 02:38:54 PM  09/02/2015 02:45:25 PM   
7           7    8         2  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8           8    9         2  05/30/2015 04:22:47 AM  05/30/2015 04:42:57 AM   
9           9   10         2  03/29/2015 01:46:20 PM  03/29/2015 01:56:56 PM   

  Store_and_fwd_flag  rate_code  Pickup_longitude  Pickup_latitude  \
0                  N          1        -73.950478        40.826733   
1                  N          1        -73.914124        40.823772   
2                  N          1        -73.952751        40.727180   
3                  N          1        -73.960274        40.720379   
4                  N          1        -73.829956        40.713699   
5                  N          1        -73.964096        40.710148   
6                  N          1        -73.830490        40.759331   
7                  N          1        -73.955406        40.721008   
8                  N          1        -73.956276        40.721710   
9                  N          1        -73.954079        40.811382   

   Dropoff_longitude    ...      Fare_amount  Extra  MTA_tax  Tip_amount  \
0         -73.971558    ...             10.0    0.0      0.5        1.00   
1         -73.975517    ...             23.0    0.5      0.5        0.00   
2         -73.971313    ...             19.5    0.5      0.5        0.00   
3         -73.956093    ...              6.5    1.0      0.5        2.00   
4         -73.823364    ...              4.0    1.0      0.5        0.00   
5         -73.994087    ...             13.5    0.5      0.5        0.00   
6         -73.843788    ...              7.0    0.0      0.5        1.56   
7         -74.005081    ...             29.0    0.5      0.5        6.06   
8         -73.906715    ...             16.0    0.5      0.5        3.46   
9         -73.945732    ...              9.0    0.0      0.5        0.00   

   Tolls_amount  Ehail_fee  Improvement_surcharge  Total_amount  Payment_type  \
0           0.0        NaN                    0.3         11.80             1   
1           0.0        NaN                    0.3         24.30             2   
2           0.0        NaN                    0.3         20.80             2   
3           0.0        NaN                    0.3         10.30             1   
4           0.0        NaN                    0.3          5.80             2   
5           0.0        NaN                    0.3         14.80             2   
6           0.0        NaN                    0.3          9.36             1   
7           0.0        NaN                    0.3         36.36             1   
8           0.0        NaN                    0.3         20.76             1   
9           0.0        NaN                    0.3          9.80             2   

   Trip_type  
0          1  
1          1  
2          1  
3          1  
4          1  
5          1  
6          1  
7          1  
8          1  
9          1  

[10 rows x 23 columns]

In [17]:
import sys
objectsize = sys.getsizeof(df2)
objectsize

56

Die größe der Datei ist unverändert, die compression existiert nur auf der Festplatte.<br>
Aber wir wissen, das Problem beim arbeiten mit großen Daten ist die Lesegeschwindigkeit der Festplatte und da helfen kleinere Dateien

<a href:'http://dask.pydata.org/en/latest/dataframe-api.html#store-dataframes'> Dask API read_csv/to_csv</a>

Mit dem folgenden Begriff kann man die Dateien zu einer großen Datei vereinen zum weitersenden etc.

In [18]:
from glob import glob
filenames = glob(r"Data\yellow_tripdata_2015-01-06longer10.*.csv")
with open(r"Data\yellow_tripdata_2015-01-06longer10Full.csv", 'w') as out:
    for fn in filenames:
        with open(fn) as f:
            out.write(f.read())

FileNotFoundError: [Errno 2] No such file or directory: 'Data\\yellow_tripdata_2015-01-06longer10Full.csv'

# Weitere Funktionen in Dask

group by Funktionen

In [25]:
df3 = df2.groupby('Payment_type').Trip_distance.mean().compute()
df3.head()

AttributeError: 'Series' object has no attribute 'compute'

In [20]:
trip_length = df2.groupby(df2['Passenger_count']).Trip_distance.max().compute()
trip_length

Passenger_count
0      4.40
1    137.30
2     47.95
3     33.63
4     42.10
5     32.30
6     21.25
7      0.00
8      2.75
9     11.90
Name: Trip_distance, dtype: float64

In [24]:
df3.to_csv(r"df3.csv", mode = 'w+')
import pandas as pd
df3 = pd.read_csv(r'df3.csv')
df3.columns  = ['Payment_type','avgtrip_distane']
df3

ValueError: Length mismatch: Expected axis has 3 elements, new values have 2 elements

Merge Funktion <br>
df3 ist ein Pandas Dataframe. Da Dask auf Pandas aufgebaut wurde sind Pandas und Dask in vielen Funktionen kompatibel

In [23]:
df2 = df2.merge(df3, how = 'inner', on='Payment_type').compute()
df2.head(10)

AttributeError: 'DataFrame' object has no attribute 'compute'

Map Funktion

In [30]:
import math
df['Pickup_longitude'] = df['Pickup_longitude'].map(lambda x: math.ceil(x) )
df['Dropoff_longitude'] = df['Dropoff_longitude'].map(lambda x: math.ceil(x) )
df.head(10)

No.  vendorid         pickup_datetime        dropoff_datetime  \
0    1         1  03/12/2015 09:26:05 AM  03/12/2015 09:33:58 AM   
1    2         1  05/29/2015 03:27:07 AM  05/29/2015 03:45:28 AM   
2    3         2  11/24/2015 08:07:46 PM  11/24/2015 08:31:35 PM   
3    4         2  12/18/2015 06:46:25 PM  12/18/2015 06:53:06 PM   
4    5         2  09/22/2015 06:32:36 PM  09/22/2015 06:34:44 PM   
5    6         2  05/15/2015 08:42:44 PM  05/15/2015 08:56:03 PM   
6    7         2  09/02/2015 02:38:54 PM  09/02/2015 02:45:25 PM   
7    8         2  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8    9         2  05/30/2015 04:22:47 AM  05/30/2015 04:42:57 AM   
9   10         2  03/29/2015 01:46:20 PM  03/29/2015 01:56:56 PM   

  Store_and_fwd_flag  rate_code  Pickup_longitude  Pickup_latitude  \
0                  N          1               -73        40.826733   
1                  N          1               -73        40.823772   
2                  N          1               -73        40.727180   
3                  N          1               -73        40.720379   
4                  N          1               -73        40.713699   
5                  N          1               -73        40.710148   
6                  N          1               -73        40.759331   
7                  N          1               -73        40.721008   
8                  N          1               -73        40.721710   
9                  N          1               -73        40.811382   

   Dropoff_longitude  Dropoff_latitude    ...      Fare_amount  Extra  \
0                -73         40.795403    ...             10.0    0.0   
1                -73         40.751713    ...             23.0    0.5   
2                -73         40.693260    ...             19.5    0.5   
3                -73         40.732990    ...              6.5    1.0   
4                -73         40.713306    ...              4.0    1.0   
5                -73         40.695229    ...             13.5    0.5   
6                -73         40.771461    ...              7.0    0.0   
7                -74         40.739941    ...             29.0    0.5   
8                -73         40.694427    ...             16.0    0.5   
9                -73         40.833076    ...              9.0    0.0   

   MTA_tax  Tip_amount  Tolls_amount  Ehail_fee  Improvement_surcharge  \
0      0.5        1.00           0.0        NaN                    0.3   
1      0.5        0.00           0.0        NaN                    0.3   
2      0.5        0.00           0.0        NaN                    0.3   
3      0.5        2.00           0.0        NaN                    0.3   
4      0.5        0.00           0.0        NaN                    0.3   
5      0.5        0.00           0.0        NaN                    0.3   
6      0.5        1.56           0.0        NaN                    0.3   
7      0.5        6.06           0.0        NaN                    0.3   
8      0.5        3.46           0.0        NaN                    0.3   
9      0.5        0.00           0.0        NaN                    0.3   

   Total_amount  Payment_type  Trip_type  
0         11.80             1          1  
1         24.30             2          1  
2         20.80             2          1  
3         10.30             1          1  
4          5.80             2          1  
5         14.80             2          1  
6          9.36             1          1  
7         36.36             1          1  
8         20.76             1          1  
9          9.80             2          1  

[10 rows x 22 columns]

Auch die where Funktion ist vorhanden.

In [32]:
df5 = df.where(df['Pickup_longitude'] != df['Dropoff_longitude'])
df5.compute()

No.  vendorid         pickup_datetime        dropoff_datetime  \
0          NaN       NaN                     NaN                     NaN   
1          NaN       NaN                     NaN                     NaN   
2          NaN       NaN                     NaN                     NaN   
3          NaN       NaN                     NaN                     NaN   
4          NaN       NaN                     NaN                     NaN   
5          NaN       NaN                     NaN                     NaN   
6          NaN       NaN                     NaN                     NaN   
7          8.0       2.0  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
8          NaN       NaN                     NaN                     NaN   
9          NaN       NaN                     NaN                     NaN   
10         NaN       NaN                     NaN                     NaN   
11         NaN       NaN                     NaN                     NaN   
12         NaN       NaN                     NaN                     NaN   
13         NaN       NaN                     NaN                     NaN   
14         NaN       NaN                     NaN                     NaN   
15         NaN       NaN                     NaN                     NaN   
16         NaN       NaN                     NaN                     NaN   
17         NaN       NaN                     NaN                     NaN   
18         NaN       NaN                     NaN                     NaN   
19         NaN       NaN                     NaN                     NaN   
20         NaN       NaN                     NaN                     NaN   
21         NaN       NaN                     NaN                     NaN   
22         NaN       NaN                     NaN                     NaN   
23         NaN       NaN                     NaN                     NaN   
24         NaN       NaN                     NaN                     NaN   
25         NaN       NaN                     NaN                     NaN   
26         NaN       NaN                     NaN                     NaN   
27         NaN       NaN                     NaN                     NaN   
28         NaN       NaN                     NaN                     NaN   
29         NaN       NaN                     NaN                     NaN   
...        ...       ...                     ...                     ...   
99970      NaN       NaN                     NaN                     NaN   
99971      NaN       NaN                     NaN                     NaN   
99972      NaN       NaN                     NaN                     NaN   
99973      NaN       NaN                     NaN                     NaN   
99974      NaN       NaN                     NaN                     NaN   
99975      NaN       NaN                     NaN                     NaN   
99976      NaN       NaN                     NaN                     NaN   
99977      NaN       NaN                     NaN                     NaN   
99978      NaN       NaN                     NaN                     NaN   
99979  99980.0       2.0  03/25/2015 10:14:34 AM  03/25/2015 10:47:26 AM   
99980      NaN       NaN                     NaN                     NaN   
99981      NaN       NaN                     NaN                     NaN   
99982      NaN       NaN                     NaN                     NaN   
99983      NaN       NaN                     NaN                     NaN   
99984      NaN       NaN                     NaN                     NaN   
99985      NaN       NaN                     NaN                     NaN   
99986      NaN       NaN                     NaN                     NaN   
99987      NaN       NaN                     NaN                     NaN   
99988      NaN       NaN                     NaN                     NaN   
99989      NaN       NaN                     NaN                     NaN   
99990      NaN       NaN                     NaN           

und auch die dropna Funktion ist vorhanden

In [33]:
df5 = df5.dropna(how = 'all')
df5.compute()

No.  vendorid         pickup_datetime        dropoff_datetime  \
7          8.0       2.0  03/29/2015 04:56:26 AM  03/29/2015 05:36:26 AM   
54        55.0       2.0  10/01/2015 02:35:59 PM  10/01/2015 03:05:13 PM   
74        75.0       2.0  07/16/2015 12:58:11 AM  07/16/2015 01:48:35 AM   
87        88.0       1.0  05/17/2015 08:17:01 PM  05/17/2015 08:45:12 PM   
98        99.0       2.0  08/20/2015 11:51:19 PM  08/21/2015 12:02:24 AM   
102      103.0       2.0  11/26/2015 08:06:53 PM  11/26/2015 08:21:48 PM   
114      115.0       2.0  09/15/2015 08:34:50 PM  09/15/2015 08:39:04 PM   
134      135.0       2.0  06/09/2015 02:14:02 PM  06/09/2015 02:54:45 PM   
149      150.0       2.0  09/06/2015 04:22:03 PM  09/06/2015 04:28:26 PM   
154      155.0       1.0  03/24/2015 11:03:32 AM  03/24/2015 11:24:17 AM   
172      173.0       2.0  10/30/2015 03:14:01 PM  10/30/2015 03:28:03 PM   
208      209.0       2.0  08/01/2015 06:16:24 PM  08/01/2015 06:32:40 PM   
211      212.0       1.0  06/02/2015 03:43:02 PM  06/02/2015 04:23:16 PM   
212      213.0       1.0  07/01/2015 11:01:58 AM  07/01/2015 11:11:18 AM   
220      221.0       2.0  11/29/2015 02:31:12 AM  11/29/2015 03:00:56 AM   
248      249.0       2.0  12/13/2015 01:39:38 AM  12/13/2015 01:45:46 AM   
249      250.0       2.0  08/21/2015 10:43:06 PM  08/21/2015 10:52:32 PM   
261      262.0       2.0  08/22/2015 03:46:55 PM  08/22/2015 03:58:34 PM   
271      272.0       2.0  04/23/2015 01:20:41 PM  04/23/2015 01:42:22 PM   
281      282.0       1.0  11/18/2015 02:36:49 PM  11/18/2015 02:48:10 PM   
294      295.0       2.0  04/26/2015 01:36:52 AM  04/26/2015 02:20:12 AM   
375      376.0       2.0  11/14/2015 12:21:40 PM  11/14/2015 12:46:58 PM   
384      385.0       2.0  03/12/2015 03:15:57 AM  03/12/2015 03:39:04 AM   
386      387.0       1.0  05/08/2015 01:20:12 PM  05/08/2015 01:44:25 PM   
410      411.0       2.0  03/22/2015 11:27:32 AM  03/22/2015 11:42:16 AM   
431      432.0       1.0  03/22/2015 04:43:45 PM  03/22/2015 04:54:19 PM   
437      438.0       2.0  07/31/2015 09:25:27 PM  07/31/2015 09:33:08 PM   
439      440.0       2.0  07/08/2015 08:24:26 PM  07/08/2015 08:39:21 PM   
460      461.0       2.0  05/30/2015 02:08:33 AM  05/30/2015 02:23:20 AM   
464      465.0       1.0  04/30/2015 02:09:07 PM  04/30/2015 02:40:45 PM   
...        ...       ...                     ...                     ...   
99192  99193.0       2.0  12/17/2015 03:56:02 PM  12/17/2015 04:05:13 PM   
99211  99212.0       2.0  03/29/2015 06:18:31 AM  03/29/2015 07:19:02 AM   
99213  99214.0       2.0  10/21/2015 06:32:58 PM  10/21/2015 06:47:12 PM   
99219  99220.0       2.0  06/18/2015 02:34:50 PM  06/18/2015 02:42:31 PM   
99224  99225.0       2.0  09/05/2015 11:26:52 PM  09/05/2015 11:50:10 PM   
99232  99233.0       2.0  03/18/2015 10:18:21 PM  03/18/2015 10:46:51 PM   
99260  99261.0       2.0  04/10/2015 08:42:54 AM  04/10/2015 09:03:32 AM   
99288  99289.0       2.0  05/12/2015 09:21:58 PM  05/12/2015 09:45:47 PM   
99302  99303.0       1.0  03/03/2015 02:02:31 PM  03/03/2015 02:37:22 PM   
99381  99382.0       1.0  07/26/2015 01:33:56 PM  07/26/2015 01:33:56 PM   
99385  99386.0       2.0  05/30/2015 10:32:18 PM  05/30/2015 10:38:24 PM   
99424  99425.0       1.0  03/02/2015 09:31:13 PM  03/02/2015 09:44:34 PM   
99437  99438.0       2.0  06/12/2015 09:21:14 PM  06/12/2015 09:59:36 PM   
99462  99463.0       2.0  10/03/2015 04:11:53 PM  10/03/2015 04:26:35 PM   
99466  99467.0       2.0  03/28/2015 08:48:00 PM  03/28/2015 09:00:17 PM   
99524  99525.0       2.0  05/04/2015 05:43:52 PM  05/04/2015 06:20:23 PM   
99548  99549.0       2.0  12/07/2015 09:54:43 AM  12/07/2015 10:21:26 AM   
99552  99553.0       1.0  12/04/2015 02:35:46 PM  12/04/2015 02:45:45 PM   
99610  99611.0       1.0  05/17/2015 05:50:32 PM  05/17/2015 06:14:05 PM   
99638  99639.0       1.0  03/14/2015 02:08:34 PM  03/14/2015 02:23:50 PM   
99649  99650.0       2.0  07/10/2015 03:45:30 PM  07/10/201

Wir haben jetzt ein Dataframe, welches alle Touren wo der 73 zum 74 Längengrad überschritten wurden enthält

In [34]:
df5.count().compute()

No.                      4497
vendorid                 4497
pickup_datetime          4497
dropoff_datetime         4497
Store_and_fwd_flag       4497
rate_code                4497
Pickup_longitude         4497
Pickup_latitude          4497
Dropoff_longitude        4497
Dropoff_latitude         4497
Passenger_count          4497
Trip_distance            4497
Fare_amount              4497
Extra                    4497
MTA_tax                  4497
Tip_amount               4497
Tolls_amount             4497
Ehail_fee                   0
Improvement_surcharge    4497
Total_amount             4497
Payment_type             4497
Trip_type                4497
dtype: int64

insgesamt 4497 Touren von 100.000 Touren

# Welche statistischen Werte gibt es?

des weiteren gibt es rolling values(rolling avg, rolling std ...) std, mean, Quantile ...

In [35]:
df = dd.read_csv(r"taxi_hunderttausend.csv",error_bad_lines=False)

Standardabweichung von Total_amount

In [36]:
std = df['Total_amount'].std().compute()
std

12.439756879724481

Rolling average aus jeweils immer 10 Zahlen gebildet

In [37]:
rolling = df['Total_amount'].rolling(10,5).mean().compute()
rolling.head(20)


0           NaN
1           NaN
2           NaN
3           NaN
4     14.600000
5     14.633333
6     13.880000
7     16.690000
8     17.142222
9     16.408000
10    17.808000
11    16.208000
12    15.040000
13    14.886000
14    14.306000
15    14.706000
16    14.400000
17    11.759000
18    10.798000
19    11.654000
Name: Total_amount, dtype: float64

Wie viele der bezahlten Beträge sind einzigartig?

In [38]:
x = df['Total_amount'].nunique()
x.compute()

2157

# weitere Features von Dask

Dask funktioniert auch auf großen Clustern und kann Grafiken zur Auslastung zeigen

In [39]:
%%HTML #2:12
<iframe width="560" height="315" src="https://www.youtube.com/embed/PAGjm4BMKlk?" frameborder="0" allowfullscreen></iframe>

In [41]:
%%html
<img src='dashboard-report-1.jpg'>